### PR TITLE
Fix regression that caused us to sync data unnecessarily.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Build is here: https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_build/results?buildId=8517588&view=results
PR: https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/503286

### CODE REVIEW

File: src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs [hunk 1]
```diff
@@ -34,6 +34,12 @@ public AssetProvider(Checksum solutionChecksum, SolutionAssetCache assetCache, I
             _serializerService = serializerService;
         }
 
+        public T GetRequiredAsset<T>(Checksum checksum)
+        {
+            Contract.ThrowIfFalse(_assetCache.TryGetAsset<T>(checksum, out var asset));
+            return asset;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs:40]: moved away from debug asserts to strict checks.

File: src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs [hunk 2]
```diff
@@ -130,7 +126,17 @@ public async ValueTask SynchronizeAssetsAsync(HashSet<Checksum> checksums, Cance
...
             using (Logger.LogBlock(FunctionId.AssetService_SynchronizeAssetsAsync, Checksum.GetChecksumsLogInfo, checksums, cancellationToken))
             {
-                var checksumsArray = checksums.ToImmutableArray();
+                using var _1 = ArrayBuilder<Checksum>.GetInstance(checksums.Count, out var missingChecksums);
+                foreach (var checksum in checksums)
+                {
+                    if (checksum == Checksum.Null)
+                        continue;
+
+                    if (!_assetCache.TryGetAsset<object>(checksum, out _))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs:135]: moved this check from callsites *into* the asset provider.  it now ensures that it doesn't request anything if it already has it (Regardless of what the caller passed in).

File: src/Workspaces/Remote/ServiceHub/Host/ChecksumSynchronizer.cs [hunk 3]
```diff
@@ -114,27 +112,21 @@ private async ValueTask CollectProjectStateChecksumsAsync(HashSet<Checksum> set,
...
-        private void AddIfNeeded(HashSet<Checksum> checksums, ChecksumCollection checksumCollection)
+        private static void AddAll(HashSet<Checksum> checksums, ChecksumCollection checksumCollection)
         {
             foreach (var checksum in checksumCollection)
-                AddIfNeeded(checksums, checksum);
-        }
-
-        private void AddIfNeeded(HashSet<Checksum> checksums, Checksum checksum)
-        {
-            if (checksum != Checksum.Null && !_assetProvider.EnsureCacheEntryIfExists(checksum))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/ChecksumSynchronizer.cs:137]: this moved into the callee

File: src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs [hunk 4]
```diff
@@ -513,9 +513,7 @@ private async Task<TextDocument> UpdateDocumentInfoAsync(TextDocument document,
 
                 foreach (var kv in documentChecksums)
                 {
-                    Debug.Assert(assetProvider.EnsureCacheEntryIfExists(kv.Item2.Info), "Expected the prior call to GetAssetsAsync to obtain all items for this loop.");
-
-                    var info = await assetProvider.GetAssetAsync<DocumentInfo.DocumentAttributes>(kv.Item2.Info, cancellationToken).ConfigureAwait(false);
+                    var info = assetProvider.GetRequiredAsset<DocumentInfo.DocumentAttributes>(kv.Item2.Info);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs:516]: this is safe, and correct (which the debug assert was alreay checking).  on line 512 we requests all these doc-attribute assets for these checksums.  So the requirement is that they *must* now be here when we go fetch them.

File: src/Workspaces/Remote/ServiceHub/Host/ChecksumSynchronizer.cs [hunk 5]
```diff
@@ -114,27 +112,21 @@ private async ValueTask CollectProjectStateChecksumsAsync(HashSet<Checksum> set,
...
+                AddAll(set, checksumObject.AnalyzerConfigDocuments);
             }
         }
 
-        private void AddIfNeeded(HashSet<Checksum> checksums, ChecksumCollection checksumCollection)
+        private static void AddAll(HashSet<Checksum> checksums, ChecksumCollection checksumCollection)
         {
             foreach (var checksum in checksumCollection)
-                AddIfNeeded(checksums, checksum);
-        }
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/ChecksumSynchronizer.cs:133]: this exists just to avoid allocations.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Review by arkalyanms — APPROVED** [review]

**Pull request merged**